### PR TITLE
[#28] [Backend] As a user, I can submit survey's answers

### PR DESCRIPTION
**Pull request by ducbm051291**
- #28 

## What happened

- The user can submit answer of the survey by calling API [Submit Survey](https://documenter.getpostman.com/view/11835486/UVsPNPfU#8e233fe5-b91e-4b98-83fc-ad3c98775180)
 
## Insight

- Update RequestConfiguration: Add Survey Response case
- Create Survey Response Parameters
- Create NoReply and APINoReply for no reply API response
- Create and implement SurveyResponseRepository
- Create SubmitSurveyResponseUseCase
- Write UnitTest for SurveyResponseRepository and SubmitSurveyResponseUseCase
- Register SurveyResponseRepository and SubmitSurveyResponseUseCase
- Update DisplayType
- Update NetworkAPI for no reply response
- Create JSONDecoder+Common
- Update related code and UnitTest

## Proof Of Work

Screenshot 2023-01-05 at 09 14 39

https://user-images.githubusercontent.com/24598204/210686373-b3cfbc7c-d26a-476d-803d-afb853a1d13e.mov

### CODE REVIEW

File: Surveys/Sources/Data/NetworkAPI/NetworkAPI.swift [hunk 1]
```diff
@@ -23,10 +28,13 @@ final class NetworkAPI: NetworkAPIProtocol {
...
     private let provider = MoyaProvider<RequestConfiguration>(plugins: [AuthPlugin()])
-    private let decoder: JSONAPIDecoder
+    private let decoder: JSONDecoder
 
-    init(decoder: JSONAPIDecoder = JSONAPIDecoder.default) {
-        self.decoder = decoder
+    init(decoderType: DecoderType) {
+        switch decoderType {
+        case .json: decoder = JSONDecoder.common
+        case .jsonAPI: decoder = JSONAPIDecoder.default
```
[the remark this answers is not in the document]
Comment (ducbm051291) [reply Surveys/Sources/Data/NetworkAPI/NetworkAPI.swift:36]: No we don't need that if we can use `.common` for all APIs. This common is only for parse empty response

**Review by Shayokh144 — APPROVED** [review]

File: Surveys/Sources/Domain/Entities/NoReply.swift [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+//  NoReply.swift
+//  Surveys
+//
+//  Created by David Bui on 04/01/2023.
+//  Copyright © 2023 Nimble. All rights reserved.
+//
+
+import Foundation
+
+protocol NoReply {}
```
Comment (blyscuit) [inline Surveys/Sources/Domain/Entities/NoReply.swift:11]: We won't need this protocol as ApiNoReply can stand on its own in the Data layer. The Domain layer can use `Void`.

File: Surveys/Sources/Data/NetworkAPI/NetworkAPI.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (blyscuit) [reply Surveys/Sources/Data/NetworkAPI/NetworkAPI.swift:36]: I found the naming for the new `json` to be lacking in detail. What do you think we rename the enum `json` to `emptyJson` which is when we use the new decoder?

**Review by blyscuit — COMMENTED** [review]
Looks good, small comments.

**Comment by blyscuit** [comment]
@ducbm051291 Why the change in https://github.com/ducbm051291/surveys-ic/pull/97/commits/85c288c9901506511a8e2d3a271dd535c94b3ac5, sending domain layer object then mapping with repository to data layer object is already correct. If we need to send more data we should create a new object in the domain layer and have it map in the data layer 🤔

**Comment by ducbm051291** [comment]
> @ducbm051291 Why the change in [85c288c](https://github.com/ducbm051291/surveys-ic/commit/85c288c9901506511a8e2d3a271dd535c94b3ac5), sending domain layer object then mapping with repository to data layer object is already correct. If we need to send more data we should create a new object in the domain layer and have it map in the data layer 🤔

They used to be parameters to request API, then I want to take advantage of them and use in question views to store the answer. That's why I renamed it to an API object. Should I keep it as before and create a new object for this?

**Comment by ducbm051291** [comment]
@blyscuit I've renamed APIQuestionResponse and APIAnswerResponse to QuestionResponse and AnswerResponse, then I move them to Domain - Entities

File: Surveys/Sources/Data/Repositories/SurveyResponseRepository.swift [hunk 3]
```diff
@@ -14,7 +14,7 @@ final class SurveyResponseRepository: SurveyResponseRepositoryProtocol {
 
     @Injected(name: .jsonNetworkAPI) private var networkAPI: NetworkAPIProtocol
 
-    func submitResponse(_ surveyId: String, questions: [APIQuestionResponse]) -> Observable<Void> {
+    func submitResponse(_ surveyId: String, questions: [QuestionResponse]) -> Observable<Void> {
```
Comment (blyscuit) [inline Surveys/Sources/Data/Repositories/SurveyResponseRepository.swift:17]: Ideally `SurveyResponseParameter` should be in data layer, this repository can map `QuestionResponse` to `SurveyResponseParameter` and `SurveyResponseParameter` should have its own `APIQuestionResponse` struct.

With current approach, the data layer `SurveyResponseParameter` is depending on the domain layer `QuestionResponse`, so it's impossible to modify `QuestionResponse` without breaking this API.

Please take a look if we are using the same object `QuestionResponse` in too many places than it should. Other than that, looks good.

File: Surveys/Sources/Domain/Entities/questionResponse.swift [hunk 4]
```diff
@@ -0,0 +1,15 @@
+//
```
[the remark this answers is not in the document]
Comment (blyscuit) [reply Surveys/Sources/Domain/Entities/questionResponse.swift:1]: Please fix danger.

File: Surveys/Sources/Data/Repositories/SurveyResponseRepository.swift (hunk 3 above)
Comment (ducbm051291) [reply Surveys/Sources/Data/Repositories/SurveyResponseRepository.swift:17]: @blyscuit I just pushed code. I created APIQuestionResponse, APIAnswerResponse for SurveyRessponseParameter

File: Surveys/Sources/Domain/Entities/questionResponse.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (ducbm051291) [reply Surveys/Sources/Domain/Entities/questionResponse.swift]: Done 👍

**Review by blyscuit — APPROVED** [review]

**Pull request merged**